File: _score/score_module.php
<?php
/**
 * Materia
 * It's a thing
 *
 * @package	    Materia
 * @version    1.0
 * @author     UCF New Media
 * @copyright  2011 New Media
 * @link       http://kogneato.com
 */


/**
 * NEEDS DOCUMENTATION
 *
 * The widget managers for the Materia package.
 *
 * @package	    Main
 * @subpackage  scoring
 * @category    Modules
  * @author      ADD NAME HERE
 */

namespace Materia;

class Score_Modules_Crossword extends Score_Module
{
	/** @var array Stores the individual scores in an array using question item_ids as keys */
	const INITIAL_SCORE = 100;

	const HINT_INTERACTION = 'question_hint';

	public $points_lost;
	public $hint_deductions;
	public $modifiers = [];
	protected $scores = [];

	protected function handle_log_question_answered($log)
	{
		$log->text = strtolower($log->text);
		$this->total_questions++;
		$this->scores[$log->item_id] = $this->check_answer($log); // score the question and add it to the total
	}

	protected function handle_log_widget_interaction($log)
	{
		if ($log->text == $this::HINT_INTERACTION)
		{
			$this->modifiers[$log->item_id] = (float) $log->value;
			// Add hit deduction to question's answer's feedback
			$this->questions[$log->item_id]->answers[0]['options']['feedback'] = 'Hint Received';
		}
	}

	// calculate the percentage and count total points
	protected function calculate_score()
	{
		$points_lost     = [];
		$updated_scores  = [];
		$hint_deductions = [];

		// if a score has a matching modifier associated with it, reduce the maximum possible question score by that amount
		foreach ($this->scores as $score_key => $score_val)
		{
			$updated_scores[$score_key] = $score_val;
			$points_lost[$score_key] = $this::INITIAL_SCORE - $score_val;

			if (isset($this->modifiers[$score_key]))
			{
				// Modifier is a negative integer, need to convert to a positive percentage
				$updated_scores[$score_key] *= 1 - (-$this->modifiers[$score_key] / $this::INITIAL_SCORE);
				$hint_deductions[$score_key] = $score_val - $updated_scores[$score_key];
			}
		}

		// Cannot generate a puzzle with zero questions => $this->total_questions > 0
		$this->points_lost        = -1 * array_sum($points_lost) / $this->total_questions;
		$this->hint_deductions    = -1 * array_sum($hint_deductions) / $this->total_questions;
		$this->verified_score     = array_sum($updated_scores);
		$this->calculated_percent = $this->verified_score / $this->total_questions;
	}

	public function check_answer($log)
	{
		if (isset($this->questions[$log->item_id]))
		{
			$question  = $this->questions[$log->item_id];
			$answer    = $this->normalize_string($question->answers[0]['text']);
			$submitted = $this->normalize_string($log->text);
			$match_num = 0;
			$guessable = 0;
			// Student could have left answer blank
			if (count($answer) > 0)
			{
				foreach ($answer as $index => $letter)
				{
					// if the letter is a guessable letter, make sure it matches
					if ($this->is_guessable_letter($letter))
					{
						$guessable++;
						if ($letter == $submitted[$index])
						{
							$match_num++;
						}
					}
				}
			}
			if ($guessable == 0) return 0;
			return $match_num / (float)$guessable * $this::INITIAL_SCORE;
		}
		return 0;
	}

	private function normalize_string($string)
	{
		return str_split(strtolower($string));
	}

	private function is_guessable_letter($char)
	{
		return  preg_match('/^[0-9a-z]+$/i', $char);
	}

	public function get_score_page_answer($log)
	{
		$question  = $this->questions[$log->item_id];
		$answer    = $this->normalize_string($question->answers[0]['text']);
		$submitted = $this->normalize_string($log->text);

		foreach ($answer as $index => $letter)
		{
			// space is guessable but user left blank, replace space with underscore
			if ($this->is_guessable_letter($letter) && $submitted[$index] == ' ')
			{
				$submitted[$index] = '_';
			}
		}

		return implode($submitted);
	}

	protected function get_feedback($log, $answers)
	{
		foreach ($answers as $answer)
		{
			if (isset($answer['options']['feedback']))
			{
				return $answer['options']['feedback'];
			}
		}
	}

	protected function get_overview_items()
	{
		$overview_items = [];

		if ($this->hint_deductions < 0)
		{
			$overview_items[] = ['message' => 'Hint Deductions', 'value' => $this->hint_deductions];
		}

		$overview_items[] = ['message' => 'Points Lost', 'value' => $this->points_lost];
		$overview_items[] = ['message' => 'Final Score', 'value' => $this->calculated_percent];

		return $overview_items;
	}
}

File: src/_score/score_module.py
from core.models import Log
from scoring.module import ScoreModule


class Crossword(ScoreModule):
    INITIAL_SCORE = 100
    HINT_INTERACTION = "question_hint"

    def __init__(self, play):
        super().__init__(play)
        self.points_lost = 0.0
        self.hint_deductions = 0.0
        self.modifiers = {}
        self.incorrect_deductions = []

    def check_answer(self, log):

        question = self.get_question_by_item_id(log.item_id)
        if not question:
            return 0

        correct_text = question.get("answers")[0]["text"]
        answer = self.normalize_string(correct_text)
        response = self.normalize_string(log.text)

        max_len = max(len(answer), len(response))

        match_num = 0
        guessable = 0

        for i in range(max_len):
            if self.is_guessable_letter(answer[i]):
                guessable += 1
                if answer[i] == response[i]:
                    match_num += 1

        if guessable == 0:
            return 0

        percent_correct = match_num / guessable
        base_score = self.INITIAL_SCORE

        # save the raw point deduction
        self.incorrect_deductions.append(
            base_score - percent_correct * base_score
        )

        # Default to full credit unless a hint modifier is present
        if log.item_id in self.modifiers:
            deduction = float(self.modifiers[log.item_id])
            hint_factor = 1 - (- deduction / self.INITIAL_SCORE)
            score = (base_score * percent_correct) * hint_factor
            self.hint_deductions += (base_score * percent_correct) - score
        else:
            score = base_score * percent_correct

        return score

    def handle_log_widget_interaction(self, log):
        if log.text == self.HINT_INTERACTION:
            self.modifiers[log.item_id] = log.value

    def calculate_score(self):
        super().calculate_score()
        if self.total_questions > 0:
            self.hint_deductions = -1 * (self.hint_deductions /
                                         self.total_questions)

            # points lost represents the point value deduction
            # from incorrect letter guesses, sans hint deduction
            self.points_lost = -1 * (
                sum(self.incorrect_deductions) / self.total_questions
            )

    def normalize_string(self, string_val: str):
        return list(string_val.lower())

    def is_guessable_letter(self, char: str) -> bool:
        return char.isalpha() or char.isdigit()

    def get_ss_answer(self, log, question):
        correct_text = question.get("answers")[0]["text"]
        answer_chars = self.normalize_string(correct_text)
        submitted_chars = self.normalize_string(log.text)

        max_len = max(len(answer_chars), len(submitted_chars))
        if len(submitted_chars) < max_len:
            submitted_chars += [" "] * (max_len - len(submitted_chars))
        for i in range(max_len):
            if (i < len(answer_chars) and 
                    self.is_guessable_letter(answer_chars[i])):
                if i < len(submitted_chars) and submitted_chars[i] == ' ':
                    submitted_chars[i] = '_'
        return "".join(submitted_chars)

    def get_feedback(self, log, answers):
        for play_log in self.logs:
            if (play_log.log_type == Log.LogType.SCORE_WIDGET_INTERACTION
                    and play_log.item_id == log.item_id):
                return "Hint Received"

        return None

    def get_overview_items(self):
        overview = []
        if self.hint_deductions < 0:
            overview.append({
                "message": "Hint Deductions",
                "value": self.hint_deductions
            })
        overview.append({
            "message": "Points Lost",
            "value": self.points_lost
            })
        overview.append({
            "message": "Final Score",
            "value": self.calculated_percent
            })
        return overview
